src/ZkFold/Cardano/UPLC/Rollup.hs

```diff
@@ -1,56 +1,120 @@
 {-# INLINABLE rollup #-}
-rollup :: SetupBytes -> RollupRedeemer -> ScriptContext -> Bool
-rollup ledgerRules (RollupRedeemer proof addr val state update) ctx =
-        -- Verify the transition from the current state to the next state
-        verify @PlonkPlutus @HaskellCore ledgerRules state' proof
-        -- Check the current rollup output
-        && out  == TxOut addr val (OutputDatum $ Datum $ toBuiltinData state)  Nothing
-        -- Check the next rollup output
-        && out' == TxOut addr val (OutputDatum $ Datum $ toBuiltinData state') Nothing
-    where
-        -- Get the current rollup output
-        Just j = findOwnInput ctx
-        out    = txInInfoResolved j
-
-        -- Get the next rollup output
-        out'   = head $ txInfoOutputs $ scriptContextTxInfo ctx
-
-        -- Compute the next state
-        state' = toInput $ dataToBlake (state, update)
+rollup :: RollupSetup -> RollupRedeemer -> ScriptContext -> Bool
+rollup (RollupSetup ledgerRules dataCurrency threadValue feeAddress) (UpdateRollup proof update) ctx =
+  let
+    -- Get the current rollup output
+    out = txInInfoResolved $ case findOwnInput ctx of
+      Just j -> j
+      _      -> traceError "rollup: no input"
+
+    -- Get the address and state of the rollup
+    (addr, val, state) = case out of
+      TxOut a v (OutputDatum (Datum s)) _ -> (a, v, unsafeDataAsI s)
+      _                                   -> traceError "rollup: invalid redeemer"
+
+    -- Get state updates as token names of the data currency
+    update' =
+      map (unTokenName . fst) $
+      concatMap (toList . snd) $
+      filter (\(k, _) -> k == dataCurrency) $
+      -- Every referenced input must have the data currency as the second currency (the first one is ada).
+      map (head . tail . toList . getValue . txOutValue . txInInfoResolved)
+      (txInfoReferenceInputs $ scriptContextTxInfo ctx)
+
+    -- Get the next rollup output
+    out'   = head $ txInfoOutputs $ scriptContextTxInfo ctx
+
+    -- Get the fee output
+    outFee = head $ tail $ txInfoOutputs $ scriptContextTxInfo ctx
+    feeVal = getValue $ txOutValue outFee
+
+    -- Get bridge outputs
+    -- If the payment credential of the output coincides with the rollup payment credential, then this output transfers value to the rollup.
+    -- Otherwise, it transfers value from the rollup.
+    bridgeOutputs =
+      filter (\(TxOut _ _ (OutputDatumHash _) Nothing) -> True) $
+      tail $ tail $ tail $ txInfoOutputs $ scriptContextTxInfo ctx
+
+    -- Compute the next state
+    state' = byteStringToInteger BigEndian $ dataToBlake (toF state, update, bridgeOutputs, feeVal)
+  in
+    -- Verify the transition from the current state to the next state
+    verify @PlonkPlutus @HaskellCore ledgerRules (toF state') proof
+
+    -- Compare the state updates
+    -- Note: we want to have the full control over the order of data updates. That is why we pass `update` in the redeemer.
+    && sort update' == sort update
```

Comment:
No, that would mean that `update` was sorted. We need `update` to be in the order we want. On the other hand, `update'` is not sorted either. So, we have to sort them both and then compare.

Reply:
By "order we want," I mean we want to determine the order when constructing the transaction instead of pre-sorting off-chain to get a particular order. This helps us to compute proofs for rollup updates more efficiently.